README.md
# Annotation API

[![Build Status](https://travis-ci.org/npm/annotation-api.svg)](https://travis-ci.org/npm/annotation-api)
[![Coverage Status](https://coveralls.io/repos/npm/annotation-api/badge.svg?branch=)](https://coveralls.io/r/npm/annotation-api?branch=master)
[![Standard Version](https://img.shields.io/badge/release-standard%20version-brightgreen.svg)](https://github.com/conventional-changelog/standard-version)

API for polling a set of external integrations and annotating packages and other entities on npmjs.com.

## License

ISC


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.6.0"></a>
# [1.6.0](https://github.com/npm/annotation-api/compare/v1.5.0...v1.6.0) (2016-06-10)


### Features

* add endpoint to support scoped packages ([#12](https://github.com/npm/annotation-api/issues/12)) ([6b97d13](https://github.com/npm/annotation-api/commit/6b97d13))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/npm/annotation-api/compare/v1.4.2...v1.5.0) (2016-05-19)


### Bug Fixes

* cache related performance tweaks ([#9](https://github.com/npm/annotation-api/issues/9)) ([392cef3](https://github.com/npm/annotation-api/commit/392cef3))
* wait for cache to populate in test ([bbc0c70](https://github.com/npm/annotation-api/commit/bbc0c70))


### Features

* include package document in outbound webhook ([#10](https://github.com/npm/annotation-api/issues/10)) ([1abc2dc](https://github.com/npm/annotation-api/commit/1abc2dc))



<a name="1.4.2"></a>
## [1.4.2](https://github.com/npm/annotation-api/compare/v1.4.1...v1.4.2) (2016-05-06)


### Bug Fixes

* when clients were removed and added, we ended up with duplicate clients in our lookup table ([#8](https://github.com/npm/annotation-api/issues/8)) ([0ecf464](https://github.com/npm/annotation-api/commit/0ecf464))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/npm/annotation-api/compare/v1.4.0...v1.4.1) (2016-05-06)


### Bug Fixes

* tweaks to better handle slow upstream services ([#6](https://github.com/npm/annotation-api/issues/6)) ([b69d68b](https://github.com/npm/annotation-api/commit/b69d68b))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/npm/annotation-api/compare/v1.3.1...v1.4.0) (2016-04-22)


### Features

* we should send along the sharedFetchSecret ([98da062](https://github.com/npm/annotation-api/commit/98da062))



<a name="1.3.1"></a>
## [1.3.1](https://github.com/npm/annotation-api/compare/v1.3.0...v1.3.1) (2016-04-22)


### Bug Fixes

* fixed a few issues found during end-to-end testing ([#5](https://github.com/npm/annotation-api/issues/5)) ([83b3705](https://github.com/npm/annotation-api/commit/83b3705)), closes [(#5](https://github.com/(/issues/5)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/npm/annotation-api/compare/v1.1.0...v1.3.0) (2016-04-19)


### Features

* add route for fetching annotations ([#4](https://github.com/npm/annotation-api/issues/4)) ([ab3a4a6](https://github.com/npm/annotation-api/commit/ab3a4a6))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/npm/annotation-api/compare/v1.1.0...v1.2.0) (2016-04-19)


### Features

* add route for fetching annotations ([2c96d3f](https://github.com/npm/annotation-api/commit/2c96d3f))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/npm/annotation-api/compare/v1.0.0...v1.1.0) (2016-04-19)


### Features

* **webhook:** adds webhook that handles fetching annotations from external integration ([#3](https://github.com/npm/annotation-api/issues/3)) ([49cd3ba](https://github.com/npm/annotation-api/commit/49cd3ba))



<a name="1.0.0"></a>
# 1.0.0 (2016-04-11)


test.js
/* global describe, it, before, after */

var server = null
var annotationApi = require('./')
var request = require('request')

require('chai').should()
global.console.info = function () {}

describe('annotation-api', function () {
  before(function (done) {
    annotationApi({
      port: 9999
    }, function (_err, _server) {
      server = _server
      return done()
    })
  })

  it('returns content', function (done) {
    request.get({
      url: 'http://0.0.0.0:9999/ping',
      json: true
    }, function (req, res, body) {
      body.response.should.equal('pong')
      return done()
    })
  })

  after(function (done) {
    server.close(function () {
      return done()
    })
  })
})


test/server.js
/* global describe, it, before, after */

var server = null
var annotationApi = require('../')
var request = require('request')

require('chai').should()
global.console.info = function () {}
global.console.error = function () {}

describe('annotation-api', function () {
  before(function (done) {
    annotationApi({
      port: 9999
    }, function (_err, _server) {
      server = _server
      return done()
    })
  })

  it('returns content', function (done) {
    request.get({
      url: 'http://0.0.0.0:9999/ping',
      json: true
    }, function (req, res, body) {
      body.response.should.equal('pong')
      return done()
    })
  })

  it('supports unscoped package', function (done) {
    request.get({
      url: 'http://0.0.0.0:9999/api/v1/annotations/lodash',
      json: true
    }, function (req, res, body) {
      res.statusCode.should.equal(200)
      body.should.be.an('array').and.be.empty
      return done()
    })
  })

  it('supports scoped package', function (done) {
    request.get({
      url: 'http://0.0.0.0:9999/api/v1/annotations/@angular/core',
      json: true
    }, function (req, res, body) {
      res.statusCode.should.equal(200)
      body.should.be.an('array').and.be.empty
      return done()
    })
  })

  after(function (done) {
    server.close(function () {
      return done()
    })
  })
})


test/package-doc.js
/* global describe, it */
var nock = require('nock')

require('chai').should()

describe('PackageDoc', function () {
  describe('get', function () {
    it('returns JSON for scoped package', function (done) {
      var pkgName = '@ben/foo'
      var registry = 'http://www.example.com'
      var secret = 'abc123'
      var PackageDoc = require('../lib/package-doc')({
        registry: registry,
        secret: secret
      })
      var getPackage = nock(registry)
        .get('/' + pkgName.replace('/', '%2f'))
        .query({
          sharedFetchSecret: secret
        })
        .reply(200, {
          'dist-tags': {latest: '1.0.0'},
          versions: {'1.0.0': {name: pkgName}}
        })
      PackageDoc.get(pkgName, function (err, doc) {
        if (err) return done(err)
        doc.name.should.equal(pkgName)
        getPackage.done()
        return done()
      })
    })

    it('gracefully handles an upstream error', function (done) {
      var pkgName = '@ben/foo'
      var registry = 'http://www.example.com'
      var secret = 'abc123'
      var PackageDoc = require('../lib/package-doc')({
        registry: registry,
        secret: secret
      })
      var getPackage = nock(registry)
        .get('/' + pkgName.replace('/', '%2f'))
        .query({
          sharedFetchSecret: secret
        })
        .reply(500)
      PackageDoc.get(pkgName, function (err, doc) {
        err.code.should.equal(500)
        getPackage.done()
        return done()
      })
    })
  })
})


index.js
var restify = require('restify')
var ClientManager = require('./lib/client-manager')

module.exports = function (opts, cb) {
  var server = restify.createServer({ name: 'annotation-api' })
  var clientManager = new ClientManager(opts)

  server.use(restify.bodyParser())
  server.use(restify.queryParser())

  server.get('/ping', function (req, res, next) {
    res.send({response: 'pong'})
    return next()
  })

  function packageHandler (req, res, next, pkgName) {
    clientManager.annotationsForPageLoad(pkgName, function (annotations) {
      res.send(annotations)
      return next()
    })
  }

  server.get('/api/v1/annotations/:pkg', function (req, res, next) {
    packageHandler(req, res, next, req.params.pkg)
  })

  server.get('/api/v1/annotations/:scope/:pkg', function (req, res, next) {
    packageHandler(req, res, next, req.params.scope + '/' + req.params.pkg)
  })

  // start listening on port.
  server.listen(opts.port, opts.host, function () {
    cb(null, server)
    console.info(server.name + ' listening at ' + server.url)
  })

  clientManager.start(function () {
    console.info('loaded addon clients')
  })
}


lib/package-doc.js
var assign = require('lodash').assign
var request = require('request')
var resolve = require('url').resolve

function PackageDoc (opts) {
  assign(this, {
    registry: process.env.FRONT_DOOR_HOST,
    secret: process.env.SHARED_FETCH_SECRET
  }, opts)
}

PackageDoc.prototype.get = function (pkgName, cb) {
  pkgName = pkgName.replace('/', '%2f')

  if (!this.registry) return cb(Error('no registry configured'))

  request.get({
    url: resolve(this.registry, pkgName),
    json: true,
    qs: {
      sharedFetchSecret: this.secret
    }
  }, function (err, res, obj) {
    if (res && res.statusCode >= 400) {
      err = Error('unexpected status = ' + res.statusCode)
      err.code = res.statusCode
    }
    if (err) {
      return cb(err)
    }

    // only return latest version.
    var latest
    if (obj['dist-tags'] && obj['dist-tags'].latest && obj.versions) {
      latest = obj.versions[obj['dist-tags'].latest]
    }

    return cb(null, latest)
  })
}

module.exports = function (opts) {
  return new PackageDoc(opts)
}


lib/client-manager.js
var _ = require('lodash')
var crypto = require('crypto')
var request = require('request')
var redis = require('redis')
var map = require('async').map

function ClientManager (opts) {
  _.assign(this, {
    clients: [],
    oauthUrl: 'http://0.0.0.0:8084/client',
    redisUrl: 'redis://0.0.0.0:6379',
    redisPrefix: '__anotation_api_',
    loadFrequency: 8000,
    loadInterval: null,
    // only make requests to external sites every 15 seconds.
    ttl: 15,
    // how long should we cache if request to upstream server failed?
    ttlFailure: 2,
    packageDoc: require('./package-doc')()
  }, opts)

  this.client = redis.createClient(this.redisUrl)
}

// loads clients from our OAuth micro-service.
// these clients each represent a single integration
// (as an example, Lift.)
ClientManager.prototype.load = function (loaded) {
  var _this = this
  request.get({
    url: this.oauthUrl,
    json: true,
    qs: {
      sharedFetchSecret: process.env.SHARED_FETCH_SECRET
    }
  }, function (err, res, body) {
    if (res && res.statusCode !== 200) {
      err = Error('unexpected status code = ' + res.statusCode)
    }
    if (err) {
      return loaded(err)
    } else {
      _this.clients = body
      return loaded()
    }
  })
}

// reload clients on an interval, this solves
// two problems:
// 1. handles annotation-api booting before oauth service.
// 2. handles new clients being added.
ClientManager.prototype.start = function (cb) {
  var _this = this
  this.loadInterval = setInterval(function () {
    _this.load(function (err) {
      if (err) console.error(err.message)
      // invoke the cb provided the
      // first time that we load clients.
      if (cb) {
        cb()
        cb = null
      }
    })
  }, this.loadFrequency)
}

ClientManager.prototype.stop = function () {
  if (this.loadInterval) clearInterval(this.loadInterval)
}

// fetch annotations from external services, returning
// from cache if possible:
// {
//  "event": "package:page-load",
//  "package": "left-pad",
//  "sender": {
//    "email": "ben@example.com"
//  }
// }
ClientManager.prototype.annotationsForPageLoad = function (pkgName, cb) {
  var _this = this
  _this.client.get('lock_' + this.key(pkgName), function (err, lock) {
    var repopulateCache = !lock

    if (err) console.error(err.message)
    _this.client.lrange(_this.key(pkgName), 0, 9999, function (err, annotations) {
      if (err) console.error(err.message)
      annotations = annotations || []

      // we always return from cache, but repopulate
      // it periodically.
      cb(annotations.map(function (annotation) {
        return JSON.parse(annotation)
      }))

      // populate the cache in the background.
      if (repopulateCache) {
        _this.fetchAnnotations(pkgName, function (annotations) {
          _this.cacheAnnotations(pkgName, annotations, function () {
            console.info('populated new annotations for', pkgName)
          })
        })
      }
    })
  })
}

// fetch annotations from all the external integrations.
ClientManager.prototype.fetchAnnotations = function (pkgName, cb) {
  var _this = this
  this.packageDoc.get(pkgName, function (err, doc) {
    if (err) console.error(err.message)
    map(_this.clients, function (client, done) {
      var payload = JSON.stringify({
        event: 'package:page-load',
        package: pkgName,
        packageDoc: doc,
        sender: {
          email: client.tokens[0].user_email
        }
      })
      var signature = _this.sign(payload, client)

      request.post({
        url: client.webhook,
        body: payload,
        timeout: 5000, // only wait 5s max for upstream response.
        headers: {
          'content-type': 'application/json',
          'npm-signature': signature
        }
      }, function (err, res, body) {
        if (res && res.statusCode >= 400) {
          err = Error('unexpected status code = ' + res.statusCode)
        }
        if (err) console.error(err.message)
        // return a blank annotation on errors.
        if (err) return done(null, null)
        if (body) {
          try {
            body = JSON.parse(body)
          } catch (err) {
            console.error(err.message)
            body = {}
          }
        }

        // the integration id is used to de-dupe.
        // the UI on the frontend.
        if (body) body.name = client.name
        body = body || {}
        body.id = client.client_id

        return done(null, body)
      })
    }, function (_err, annotations) {
      return cb(annotations.filter(function (item) {
        return !!item
      }))
    })
  })
}

/*
Cache a set of annotations, so we don't hammer
the external API.

Reminder that annotations look something like this:
[{
  id: 'abc-123-abc',
  fingerprint: 'a',
  rows: [{
    image: {
      url: 'http://www.example.com/img',
      text: 'image alt'
    },
    link: {
      url: 'http://www.example.com',
      text: 'my awesome link'
    },
    text: 'hello *world*!'
  }]
}]
*/
ClientManager.prototype.cacheAnnotations = function (pkgName, annotations, cb) {
  var failures = this.clients.length !== annotations.length
  if (!annotations.length) return cb()
  // toss a fingerprint on our annotation so that the
  // frontend can dedupe and turn it into a string for Redis.
  annotations = annotations.map(function (annotation) {
    annotation.fingerprint = hash(JSON.stringify(annotation), '')
    return JSON.stringify(annotation)
  })
  this.client.setex('lock_' + this.key(pkgName), failures ? this.ttlFailure : this.ttl, true)
  this.client.del(this.key(pkgName))
  this.client.lpush(this.key(pkgName), annotations, function (err) {
    if (err) console.error(err.message)
    return cb()
  })
}

function hash (payload, secret) {
  return crypto.createHmac('sha256', secret)
    .update(payload)
    .digest('hex')
}

// helpers for signing the webhook payload.
ClientManager.prototype.sign = function (payload, client) {
  var secret = client.tokens[0].access_token
  return 'sha256=' + hash(payload, secret)
}
ClientManager.prototype.key = function (pkgName) {
  return this.redisPrefix + pkgName
}

module.exports = ClientManager
